### PR TITLE
types, plan: return ErrInvalidDefault when date type column option of create table statement has a now() default value

### PR DESCRIPTION
**Pull request by winkyao**
for #4429

### CODE REVIEW

**Comment by winkyao** [comment]
@breeswish @shenli @XuHuaiyu PTAL

File: plan/validator.go [hunk 1]
```diff
@@ -444,12 +444,27 @@ func checkColumn(colDef *ast.ColumnDef) error {
...
+				return types.ErrInvalidDefault.GenByArgs(colDef.Name.Name.O)
+			}
+		}
 	default:
 		// TODO: Add more types.
 	}
 	return nil
 }
 
+func isNowBuiltinFunc(expr ast.ExprNode) bool {
```
Comment (coocood) [inline plan/validator.go]: This function checks `now` symbolic, it doesn't have to be exactly `now`.
Use `isNowSymFunc` is better, like the one defined in parser.

File: plan/validator.go [hunk 2]
```diff
@@ -444,12 +444,27 @@ func checkColumn(colDef *ast.ColumnDef) error {
 				return types.ErrIllegalValueForType.GenByArgs(types.TypeStr(tp.Tp), str)
 			}
 		}
+	case mysql.TypeDate:
```
Comment (coocood) [inline plan/validator.go]: `NOW` function is invalid for other types too, it's valid for only `DATETIME` and `TIMESTAMP`.

**Comment by winkyao** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

File: util/types/errors.go [hunk 3]
```diff
@@ -45,6 +45,8 @@ var (
 	ErrCastAsSignedOverflow = terror.ClassTypes.New(codeUnknown, msgCastAsSignedOverflow)
 	// ErrCastNegIntAsUnsigned is returned when a negative integer be casted to an unsigned int.
 	ErrCastNegIntAsUnsigned = terror.ClassTypes.New(codeUnknown, msgCastNegIntAsUnsigned)
+	// ErrInvalidDefault is return when meet a invalid default value.
```
Comment (XuHuaiyu) [inline util/types/errors.go]: s/ return/ returned

File: plan/validator.go [hunk 4]
```diff
@@ -450,6 +454,28 @@ func checkColumn(colDef *ast.ColumnDef) error {
 	return nil
 }
 
+func isNowSymFunc(expr ast.ExprNode) bool {
```
Comment (XuHuaiyu) [inline plan/validator.go]: add a comment for this func.

File: plan/validator.go [hunk 5]
```diff
@@ -450,6 +454,28 @@ func checkColumn(colDef *ast.ColumnDef) error {
...
+	if funcCall, ok := expr.(*ast.FuncCallExpr); ok {
+		if funcCall.FnName.L == ast.CurrentTimestamp {
```
Comment (XuHuaiyu) [inline plan/validator.go]: why not check ast.Now here?

Comment (winkyao) [reply plan/validator.go]:
```
DefaultValueExpr:
	NowSymOptionFraction | SignedLiteral

NowSymOptionFraction:
	NowSym
	{
		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr("CURRENT_TIMESTAMP")}
	}
|	NowSymFunc '(' ')'
	{
		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr("CURRENT_TIMESTAMP")}
	}
|	NowSymFunc '(' NUM ')'
	{
		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr("CURRENT_TIMESTAMP")}
	}
```

@XuHuaiyu

Comment (zz-jason) [reply plan/validator.go]: we can add a comment for `isNowSymFunc` to state that default value `NOW()` is transformed to `CURRENT_TIMESTAMP()`

Comment (XuHuaiyu) [reply plan/validator.go]: ok

File: plan/validator.go [hunk 6]
```diff
@@ -450,6 +454,33 @@ func checkColumn(colDef *ast.ColumnDef) error {
 	return nil
 }
 
+// isNowSymFunc check whether defaul value is a NOW() builtin function.
```
Comment (XuHuaiyu) [inline plan/validator.go]: s/ check/ checks

**Pull request merged**